server/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java

```diff
@@ -85,85 +84,138 @@ protected ClusterBlockException checkBlock(RolloverRequest request, ClusterState
+                final Map<String, Boolean> preConditionResults = evaluateConditions(rolloverRequest.getConditions().values(),
+                    metadata.index(preSourceIndexName), statsResponse);
+
+                // If this is a dry run, return with the results without invoking a cluster state update
+                if (rolloverRequest.isDryRun()) {
+                    listener.onResponse(new RolloverResponse(preSourceIndexName, preRolloverIndexName,
+                        preConditionResults, true, false, false, false));
+                    return;
+                }
+
+                // Holders for what our final source and rolled over index names are as well as the
+                // conditions met to cause the rollover, these are needed so we wait on and report
+                // the correct indices and conditions in the clusterStateProcessed method
+                final SetOnce<String> sourceIndex = new SetOnce<>();
+                final SetOnce<String> rolloverIndex = new SetOnce<>();
+                final SetOnce<Map<String, Boolean>> conditionResults = new SetOnce<>();
+
+                final List<Condition<?>> preMetConditions = rolloverRequest.getConditions().values().stream()
+                    .filter(condition -> preConditionResults.get(condition.toString())).collect(Collectors.toList());
+
+                // Pre-check the conditions to see whether we should submit a new cluster state task
+                if (preConditionResults.size() == 0 || preMetConditions.size() > 0) {
+
+                    // Submit the cluster state, this can be thought of as a "synchronized"
+                    // block in that it is single-threaded on the master node
+                    clusterService.submitStateUpdateTask("rollover_index source [" + preRolloverIndexName + "] to target ["
+                        + preRolloverIndexName + "]", new ClusterStateUpdateTask() {
+                        @Override
+                        public ClusterState execute(ClusterState currentState) throws Exception {
+                            // Regenerate the rollover names, as a rollover could have happened
+                            // in between the pre-check and the cluster state update
+                            final MetadataRolloverService.NameResolution rolloverNames =
+                                rolloverService.resolveRolloverNames(currentState, rolloverRequest.getRolloverTarget(),
+                                    rolloverRequest.getNewIndexName(), rolloverRequest.getCreateIndexRequest());
+                            final String sourceIndexName = rolloverNames.sourceName;
+
+                            // Re-evaluate the conditions, now with our final source index name
+                            final Map<String, Boolean> postConditionResults = evaluateConditions(rolloverRequest.getConditions().values(),
+                                metadata.index(sourceIndexName), statsResponse);
```

Comment:
I'm a bit worried about the semantics of the `Multiple concurrent rollovers with conditions should roll over exactly once, assuming the condition is met only once` invariant. Namely about the `assuming the condition is met only once` bit. I think a concurrent rollover will always be rejected. 

Let's assume the conditions would be met even if there was a concurrent rollover (there was a rollover and a  document indexed when the condition was `max_docs: 1` so by the time we got to re-evaluate the conditions as part of the "rollover cluster state change" the alias/ds target is indeed ready to rollover again). 
But we are evaluating against the old `statsResponse`. The rollover will never meet the condition as the stats don't contain the now latest index - the `sourceIndex` is now generation `000002` (as the concurrent rollover created this index) but we grabbed the stats when this index didn't exist (the latest generation was `000001` in the beginning). 

I believe we should redefine the invariant here (and the implementation should we decide to change it).

Reply:
Thinking about this some more, I think it's rather unlikely for very fast consecutive rollovers to be what the users want, so I'd say if we detect the source name changed (ie. the target was already rolled over) we should not do anything (ie. ` return currentState`). The client listener will have the option to retry themselves as the API will reflect that rollover didn't happen.
What do you think?

Reply:
I agree, I glossed over the fact that the stats are only generated once per rollover (rather than invoked twice), but it is possible for the retrieval of stats to be separate from the evaluation of those conditions within the cluster state update thread. I think this is a reasonable compromise.

Reply:
> I think it's rather unlikely for very fast consecutive rollovers to be what the users want, so I'd say if we detect the source name changed (ie. the target was already rolled over) we should not do anything (ie. return currentState).

I think I'd prefer to err on the side of the behavior being consistent with regard to the number of requests, for example, N unconditional rollover requests should roll over N times. I think we (ES) shouldn't make judgements about "oh there are too many rollovers within an amount of time so I'm not going to do the subsequent ones", especially since it's possible for the rollover to be delayed (in extreme cases, it could be a while between the actual rollovers even if they are initiated at the same time)